broker-daemon/interchain/index.js

```diff
@@ -0,0 +1,275 @@
+ * @param {Payment} inboundPayment
+ * @param {Engine}  inboundPayment.engine - Engine of the expected inbound
+ *                                          payment
+ * @param {number}  inboundPayment.amount - Amount, in the smallest unit, of the
+ *                                          expected inbound payment.
+ * @param {Date}    timeout               - Absolute time after which the
+ *                                          payment should not be translated.
+ */
+async function prepareSwap (hash, { engine, amount }, timeout) {
+  // TODO: update `prepareSwap` to take an expiration
+  // time (absolute) and a minimum time lock in seconds
+  // TODO: update `prepareSwap` to be idempotent
+  await engine.prepareSwap(
+    hash,
+    amount,
+    INBOUND_TIME_LOCK,
+    timeout
+  )
+}
+
+/**
+ * The result of retrieving a preimage by paying a downstream payment locked
+ * with the swap hash.
+ * @typedef {Object} PaymentResult
+ * @property {string} paymentPreimage - Base64 encoded preimage for the swap.
+ * @property {string} paymentError    - Error message from a failed attempt to
+ *                                      translate a swap via payment.
+ *
+ */
+
+/**
+ * Get the preimage for a swap from one of its three potential sources:
+ * - from an existing outbound payment
+ * - from an existing settled inbound invoice
+ * - from a new outbound payment
+ *
+ * @private
+ *
+ * @param {string}  hash                    - Base64 string of the hash for the
+ *                                            swap
+ * @param {Payment} inboundPayment
+ * @param {Engine}  inboundPayment.engine   - Engine of the expected inbound
+ *                                            payment
+ * @param {Payment} outboundPayment
+ * @param {Engine}  outboundPayment.engine  - Engine of the outbound
+ *                                            payment
+ * @param {number}  outboundPayment.amount  - Amount, in the smallest unit, of
+ *                                            the outbound payment to be made.
+ * @param {number}  outboundPayment.address - Address to send the outbound
+ *                                            payment to to retrieve the
+ *                                            preimage.
+ * @returns {PaymentResult}                   Result of retrieving the preimage.
+ */
+async function getPreimage (
+  hash,
+  { engine: inboundEngine },
+  { engine: outboundEngine, amount: outboundAmount, address: outboundAddress }
+) {
+  // We don't want to reject an incoming HTLC if we know that there is an
+  // active outgoing one. If the outgoing HTLC is in flight or completed, we
+  // should attempt to retrieve the associated preimage.
+  logger.debug(`Checking outbound HTLC status for swap ${hash}`)
+  if (await outboundEngine.isPaymentPendingOrComplete(hash)) {
+    // TODO: return permanent errors when encountered
+    return {
+      paymentPreimage: await outboundEngine.getPaymentPreimage(hash)
```

Comment:
Should we just return the preimage rather than putting it in an object? That would make it more consistent with the function name.

Reply:
This makes it consistent with the other return values. If we don't do it like this, we have to handle the `permanentError` case inside this function, which I don't think is desirable.

Reply:
I think you could do it without affecting the error handling, something like this:
`const {preimage} = outboundEngine.translateSwap(...)`
`return preimage`

Reply:
You have to handle the `permanentError` case, like we do here: https://github.com/sparkswap/broker/pull/547/files#diff-cc53b1d80fb0169112cec8eb19c4fa7aR214

Reply:
Oh, I see, `permanentError` is an object property, I was thinking it was an exception. So why isn't it desirable to convert to an exception in this function? It seems cleaner that way from what I can see.

Reply:
I expect that it will happen in two places. Here (when this is replaced with `TrackPayment`) `outboundEngine.translateSwap`.

Reply:
Ah ok, could do a helper function to remove the duplication, but not a big deal either way